# Customer Churn Prediction using Logistic Regression

In [ ]:
# import libraries
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix, precision_score, recall_score

In [ ]:
# load dataset
df = pd.read_excel('Telco Churn dataset.xlsx')

In [ ]:
# Preview data
df.head()

,customerID,gender,SeniorCitizen,MaritalStatus,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,TotalEveMinutes,TotalEveCalls,TotalNightMinutes,TotalNightCalls,TotalIntlMinutes,TotalIntlCalls,CustomerServiceCalls,TotalCall,TotalRevenue,Churn
0,0002-ORFBO,Female,0,Yes,Yes,9,Yes,No,DSL,No,...,241.4,107,204.8,106,15.5,4,0,354,593.3,No
1,0004-TLHLJ,Male,0,No,No,4,Yes,No,Fiber optic,No,...,131.7,94,169.5,106,10.3,9,5,326,280.85,Yes
2,0013-MHZWF,Female,0,No,Yes,9,Yes,No,DSL,No,...,178.6,102,126.8,82,8.0,4,2,324,571.45,No
3,0013-SMEOE,Female,1,Yes,No,71,Yes,No,Fiber optic,Yes,...,262.2,64,130.0,92,8.8,4,0,228,7904.25,No
4,0015-UOCOJ,Female,1,No,No,7,Yes,No,DSL,Yes,...,195.7,86,236.5,137,12.0,2,1,294,340.35,No


In [ ]:
# check for missing values
df.isnull().sum()

,0
customerID,0
gender,0
SeniorCitizen,0
MaritalStatus,0
Dependents,0
tenure,0
PhoneService,0
MultipleLines,0
InternetService,0
OnlineSecurity,0


In [ ]:
# Convert the target variable to numeric
df['Churn'] = df['Churn'].map({'Yes':1, 'No':0}) # Convert categorical labels into binary values


In [ ]:
# One Hot encode categorical variables
df = pd.get_dummies(df, drop_first=True) # Convert categorical columns into numeric columns

In [ ]:
# Handle missing values
df = df.dropna()

# Train/Test Split

In [ ]:
# Define features and target
X = df.drop('Churn', axis=1) # Independent variables/Predictor
# Target Variable
y = df['Churn']

In [ ]:
# Split the dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
# 80% training, 20% testing

# Model Building

In [ ]:
# initialize model
model = LogisticRegression(max_iter=1000)

In [ ]:
# train model
model.fit(X_train, y_train) # Learn relationship between features and churn

/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(max_iter=1000)

# Predictions


In [ ]:
# predict probabilities
y_prob = model.predict_proba(X_test)[:,1] # returns probability of class 1(churn)

# convert probabilities to class labels
y_pred = (y_prob >= 0.3 ).astype(int)
# Threshold = 0.3
# if probability >= 0.3 -> churn

# Model Evaluation

In [ ]:
# Accuracy
accuracy = accuracy_score(y_test, y_pred)
accuracy


0.8830584707646177

89.65% is the percentage of correct predictions

In [ ]:
# Confusion Matrix
cm = confusion_matrix(y_test, y_pred)
cm


array([[523,  55],
       [ 23,  66]])

0          1

TN     FP

FN     TP

523 are true negative

23 are false negative

55 False positive

66 True positive

In [ ]:
# Precision
precision = precision_score(y_test, y_pred)
precision

0.5454545454545454

Precision = TP/(TP+FP)

54.5% Percent of the customers predicted to churn did indeed churn.

In [ ]:
# Recall
recall = recall_score(y_test, y_pred)
recall
# Of actual churners, how many we correctly identified

0.7415730337078652

Recall = TP/(TP+FN)

Our model has a recall of 74.16% which is quite good and this means catcching about 3 out 4 churners

# Interpretation

In [ ]:
coefficients = pd.DataFrame({
    'Feature': X.columns,
    'Coefficient': model.coef_[0]
})
coefficients.head(20)

,Feature,Coefficient
0,SeniorCitizen,0.147802
1,tenure,-0.044330
2,NumbervMailMessages,0.015108
3,TotalDayMinutes,0.010523
4,TotalDayCalls,-0.081865
5,TotalEveMinutes,0.003413
6,TotalEveCalls,-0.086405
7,TotalNightMinutes,0.000968
8,TotalNightCalls,-0.085438
9,TotalIntlMinutes,0.071464


Positive coefficient - Increase the churn probability
Negative Coefficient - decreases churn probability

CustomerServiceCalls - Higher chance of churn

Long Tenure - Lower chance of churn

# Conclusion

1. The model improved in the right direction

Lowering the threshold significantly increased Recall (43% → 74.15%), meaning the model now identifies most churners.

2. The drop in precision is expected and acceptable

Precision fell (67.2% → 54%) because the model flags more customers as churners.
This introduces more false positives, which is a normal trade-off.

3. Accuracy is not the decision metric

Accuracy only dropped slightly (89% → 88%), confirming it was never the key issue.
The earlier model looked good on accuracy but performed poorly on the actual goal.

4. The new model is better for churn prediction

Catching churners is more important than avoiding false alarms.
This version aligns better with business needs.

5. The model is now more proactive
Before: Reactive (missed many churners)
Now: Proactive (identifies most churn risks)
6. There is now a cost trade-off
More customers will be targeted unnecessarily
But fewer real churners will be lost

The model is now suitable if the cost of intervention is lower than the cost of losing a customer.

7. Final takeaway

The improved model sacrifices precision to gain recall, making it more effective for identifying churn and supporting better business decisions.